src/librender/scene.cpp

```diff
@@ -172,43 +172,117 @@ Scene<Float, Spectrum>::ray_test(const Ray3f &ray, Mask active) const {
+    // Account for the discrete probability of sampling this emitter.
+    ray_weight *= wav_weight * emitter_weight;
+#else
+    auto [ray, ray_weight] =
+        emitter->sample_ray(time, sample1, sample2, sample3, active);
+#endif
+
+    return std::make_tuple(ray, ray_weight, emitter);
+}
+
+MTS_VARIANT std::pair<typename Scene<Float, Spectrum>::UInt32, Float>
+Scene<Float, Spectrum>::sample_emitter(Float index_sample, Mask active) const {
+    MTS_MASKED_FUNCTION(ProfilerPhase::SampleEmitter, active);
+
+    size_t emitters_size = m_emitters.size();
+    // Fast paths if there is less than two emitters
+    if (unlikely(emitters_size == 0))
+        return { UInt32(-1), 0.f };
+    if (emitters_size == 1)
+        return { 0, 1 };
+
+    // Randomly pick an emitter
+#if 0
+    auto [index, emitter_pdf] = m_emitter_distr.sample_pmf(index_sample);
+    return { index, ek::rcp(emitter_pdf) };
+#else
+    UInt32 index = ek::min(UInt32(index_sample * (ScalarFloat) emitters_size),
+                           (uint32_t) emitters_size - 1);
+    // pdf = 1 / emitters_size  =>  sampling_weight = emitters_size
+    return { index, emitters_size };
+#endif
+}
+
+MTS_VARIANT std::tuple<typename Scene<Float, Spectrum>::UInt32, Float, Float>
+Scene<Float, Spectrum>::sample_emitter_reuse(Float index_sample,
+                                             Mask active) const {
+    MTS_MASKED_FUNCTION(ProfilerPhase::SampleEmitter, active);
+
+    size_t emitters_size = m_emitters.size();
+    // Fast paths if there is less than two emitters
+    if (unlikely(emitters_size == 0))
+        return { UInt32(-1), 0.f, index_sample };
+    if (emitters_size == 1)
+        return { 0, 1, index_sample };
+
+#if 0
+    // Randomly pick an emitter
+    auto [index, emitter_pdf, sample_re] = m_emitter_distr.sample_reuse_pmf(index_sample);
+    return { index, ek::rcp(emitter_pdf), sample_re };
+#else
+    UInt32 index = ek::min(UInt32(index_sample * (ScalarFloat) emitters_size),
+                           (uint32_t) emitters_size - 1);
+    // Rescale sample to lie in [0,1) again
+    index_sample = (index_sample - index / (ScalarFloat) emitters_size) * emitters_size;
+    // pdf = 1 / emitters_size  =>  sampling_weight = emitters_size
+    return { index, emitters_size, index_sample };
+#endif
+}
+
 MTS_VARIANT std::pair<typename Scene<Float, Spectrum>::DirectionSample3f, Spectrum>
 Scene<Float, Spectrum>::sample_emitter_direction(const Interaction3f &ref, const Point2f &sample_,
                                                  bool test_visibility, Mask active) const {
     MTS_MASKED_FUNCTION(ProfilerPhase::SampleEmitterDirection, active);
-
     using EmitterPtr = ek::replace_scalar_t<Float, Emitter*>;
 
     Point2f sample(sample_);
     DirectionSample3f ds;
     Spectrum spec;
 
     size_t emitters_size = m_emitters.size();
-
     if (likely(emitters_size > 0)) {
         if (emitters_size == 1) {
```

Comment:
While you are changing this function, you can remove the `emitters_size == 1` special case. Enoki now performs the 1-instance-only optimization automatically internally.